### PR TITLE
[core] Support auth in REST Catalog

### PR DESCRIPTION
**Pull request by jerry-024**
### Purpose
Support auth in REST Catalog
Linked issue: https://github.com/apache/paimon/issues/4540

### Tests

- AuthSessionTest
- CredentialsProviderFactoryTest

### API and Format
No

### Documentation
No

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+package org.apache.paimon.rest.auth;
+
+import org.apache.paimon.options.ConfigOption;
+import org.apache.paimon.options.ConfigOptions;
+
+import java.time.Duration;
+
+/** Options for REST Catalog Auth. */
+public class AuthOptions {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java:27]: It is better to move all options to `RESTCatalogOptions`.

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    public static final ConfigOption<String> TOKEN =
+            ConfigOptions.key("token")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("REST Catalog auth token.");
+    public static final ConfigOption<Duration> TOKEN_EXPIRES_IN =
+            ConfigOptions.key("token-expires-in")
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java:34]: token.expiration-time

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    .durationType()
+                    .defaultValue(Duration.ofHours(1))
+                    .withDescription("REST Catalog auth token expires in.");
+    public static final ConfigOption<Boolean> TOKEN_REFRESH_ENABLED =
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java:38]: Do we really need this option? Maybe just remove this.

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            ConfigOptions.key("token-refresh-enabled")
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription("REST Catalog auth token refresh enable.");
+    public static final ConfigOption<String> TOKEN_FILE_PATH =
+            ConfigOptions.key("token-file-path")
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java:44]: `token.provider.path`?

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("REST Catalog auth token file path.");
+    public static final ConfigOption<String> CREDENTIALS_PROVIDER =
+            ConfigOptions.key("credentials_provider")
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java:49]: It seems we don't need to provide this now, we can infer the provider from `token.provider.path` option.

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/BearTokenFileCredentialsProvider.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Override
+    public Optional<Long> expiresInMills() {
+        return Optional.ofNullable(this.expiresInMills);
+    }
+
+    private String getTokenFromFile() {
+        try {
+            // todo: handle exception
+            return new String(Files.readAllBytes(Paths.get(tokenFilePath)), StandardCharsets.UTF_8);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/BearTokenFileCredentialsProvider.java:101]: `FileIOUtils.readFileUtf8`

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/BearTokenFileCredentialsProvider.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Override
+    public Optional<Long> expiresInMills() {
+        return Optional.ofNullable(this.expiresInMills);
+    }
+
+    private String getTokenFromFile() {
+        try {
+            // todo: handle exception
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/BearTokenFileCredentialsProvider.java:100]: remove this todo.

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/BearTokenFileCredentialsProvider.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Override
+    String token() {
+        return this.token;
+    }
+
+    @Override
+    public boolean refresh() {
+        long start = System.currentTimeMillis();
+        this.token = getTokenFromFile();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/auth/BearTokenFileCredentialsProvider.java:60]: Do not update this field if refresh fails?

File: paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java (hunk 5 above)
Comment (jerry-024) [reply paimon-core/src/main/java/org/apache/paimon/rest/auth/AuthOptions.java]: Move to RESTCatalogInternalOptions, if user define this, we will it.

File: paimon-common/src/main/java/org/apache/paimon/utils/ThreadPoolUtils.java [hunk 9]
```diff
@@ -76,6 +79,13 @@ public static ThreadPoolExecutor createCachedThreadPool(
         return executor;
     }
 
+    public static ScheduledExecutorService createScheduledThreadPool(
+            int threadNum, String namePrefix) {
+        return new ScheduledThreadPoolExecutor(
+                threadNum,
+                new ThreadFactoryBuilder().setDaemon(true).setNameFormat(namePrefix).build());
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/utils/ThreadPoolUtils.java:86]: newDaemonThreadFactory?

File: paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java [hunk 10]
```diff
@@ -27,37 +26,42 @@
...
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static org.apache.paimon.utils.ThreadPoolUtils.createScheduledThreadPool;
 
 /** A catalog implementation for REST. */
 public class RESTCatalog implements Catalog {
     private RESTClient client;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java:50]: close this in catalog.close?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**